### ISSUE TITLE
Make the coregistration approaches more modular.

### ISSUE THREAD

**Issue by erikmannerfelt** [GlacioHack/xdem#24]
There is no consensus on the optimal coregistration pipeline, so adding modularity would help explore this.

### A structure suggestion

It could be done in scikit-learn's fashion for regression models:

```python
from sklearn.linear_model import LinearRegressor, RANSACRegressor
import numpy as np

X = np.linspace(0, 1, num=50).reshape(-1, 1)
y = np.linspace(0, 2, num=50)

for regressor in LinearRegressor, RANSACRegressor:
    model = regressor()
    model.fit(X, y)
    print(model.score(X, y))
```
The point of the above example is to show how interchangeable the different regression models are.

The function `sklearn.pipeline.make_pipeline()` allows for custom regression pipelines to be created. If for example the samples should be normalized before, this could easily be done:
```python
from sklearn.linear_model import LinearRegressor
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline

model = make_pipeline([StandardScaler(), LinearRegressor(fit_intercept=True)])

model.fit(X,y)
# etc...
```

A similar approach could be done for `coreg.py`:

```python
from xdem import coreg

reference_dem = ....
dem_to_be_aligned = ....
other_dem = ....

model = coreg.AmauryCoreg()
model.fit(reference_dem, dem_to_be_aligned)

print(model.nmad)

model.transform(other_dem)
```
... or something similar.
If a custom pipeline is sought:

```python
model = coreg.make_pipeline([coreg.BiasCorr(), coreg.OutlierRemoval(), coreg.ICPCoreg()])

model.fit(reference_dem, dem_to_be_aligned)
```

Basically, all coreg approaches should be subclasses of a `BaseCoreg` class or similar, and all should have the methods `.fit()` and `.transform()` (or similar terminology). The `make_pipeline()` could be adapted from the equivalent in scikit-learn.

**Comment by adehecq** [comment]
I like it!

**Comment by rhugonnet** [comment]
Yes, this looks really great!! I'd be curious of what @fmaussion thinks about this. How is the modular approach of OGGM conceived? I'm not familiar with it myself.

**Comment by adehecq** [comment]
Assuming we have the worse DEM possible, with gaps, noise and with a very large transformation, it would be good to have the following steps:
- remove a vertical shift, or possibly a ramp
- calculate residuals to a reference elevation and remove large outliers
- calculate a rigid (or possibly non rigid) transformation
- refine with Nuth & Kaab which is always more accurate on subpixel shifts
